src/functions/avl-update-endpoint/index.ts

```diff
@@ -0,0 +1,120 @@
+        }
+
+        const subscriptionId = event.pathParameters?.subscriptionId;
+
+        if (!subscriptionId) {
+            throw new Error("Subscription ID must be passed as a path parameter");
+        }
+
+        if (!event.body) {
+            throw new Error("No body sent with event");
+        }
+
+        const parsedBody = avlUpdateBodySchema.safeParse(JSON.parse(event.body));
+
+        if (!parsedBody.success) {
+            logger.error(JSON.stringify(parsedBody.error));
+            throw new Error("Invalid event body for updating subscription.");
+        }
+
+        const updateBody = parsedBody.data;
+        const { username, password } = updateBody;
+
+        const subscription = await getSubscription(subscriptionId, tableName);
+
+        if (!subscription) {
+            logger.info(`Subscription with ID: ${subscriptionId} not found in subscription table`);
+            return { statusCode: 404 };
+        }
+
+        logger.info(`Starting lambda to update subscription with ID: ${subscriptionId}`);
+
+        const subscriptionDetail: Omit<AvlSubscription, "PK" | "status"> = {
+            url: updateBody.dataProducerEndpoint,
+            description: updateBody.description ?? subscription.description,
+            shortDescription: updateBody.shortDescription ?? subscription.shortDescription,
+            requestorRef: subscription.requestorRef,
+            publisherId: subscription.publisherId,
+            serviceStartDatetime: subscription.serviceStartDatetime,
+            lastModifiedDateTime: subscription.lastModifiedDateTime ?? null,
+        };
+
+        try {
+            logger.info(`Unsubscribing from subscription ID: ${subscriptionId} using existing credentials `);
+            await sendTerminateSubscriptionRequestAndUpdateDynamo(subscriptionId, subscriptionDetail, tableName);
+        } catch (e) {
+            logger.warn(
+                `An error occurred when trying to unsubscribe from subscription with ID: ${subscriptionId}. Error ${e}`,
+            );
+        }
+
+        await addSubscriptionAuthCredsToSsm(subscriptionId, updateBody.username, updateBody.password);
+
+        logger.info(`Subscribing to subscription ID: ${subscriptionId} using new details`);
+        await sendSubscriptionRequestAndUpdateDynamo(
+            subscriptionId,
+            subscriptionDetail,
+            username,
+            password,
+            tableName,
+            dataEndpoint,
+            mockProducerSubscribeEndpoint,
+        );
+
+        logger.info(`Successfully updated subscription ID: ${subscriptionId}`);
+        return {
+            statusCode: 204,
+        };
+    } catch (e) {
+        if (e instanceof Error) {
+            logger.error(
+                `An error occurred when updating subscription ID: ${event.pathParameters?.subscriptionId ?? ""}`,
+                e,
+            );
+        }
+
+        throw e;
```

Comment:
We don't want to expose the internals of errors with HTTP lambdas so this should caught and instead be our own defined 500 response